docs/packaging-your-sql-project/intro-to-etl-pipelines-with-python-and-sql.md

```diff
@@ -11,4 +11,186 @@ kernelspec:
+2. Use `pandas` to appropriately transform the data for later use, documentation [found here](https://pandas.pydata.org/docs/)
+3. Use `SQLAlchemy` to load the data into a DuckDB file, documentation [found here](https://docs.sqlalchemy.org/en/20/)
+
+In fact, all of these steps are already included in a single script: `datadownload.py`. This script can be found from the course's GitHub repo that you can clone from here:
+
+`git clone https://github.com/ploomber/sql.git`
+
+Once you clone the repo, make sure that you have the correct dependencies by following the "Setup" instructions under the `CONTRIBUTING.md` files. Once you have the correct dependencies and environment, navigate to the `pipeline` folder and run the script with this line in your terminal: `python src/datadownload.py`. The script will extract, load, and transform the data and output a DuckDB file in the current folder with the name `car_data.duckdb`.
+
+Before we dive into understanding how this script fundamentally follows an ETL pipeline, we have to briefly expand on some key points regarding DuckDB and `pandas`:
+
+1. We use DuckDB because it allows us to store data in an in-memory database. Typically, the "loading" part of ETLs would require a server that hosts a database and usually requires additional configurations. DuckDB disregards the need for servers and allows users to locally load data. In our case, you can think of `car_data.duckdb` as our local database that we will be running queries against. DuckDB is an efficient and easy to set up database that is especially useful the in initial stages of the data science workflow. This database allows data scientists to check the usefulness of the data without having to cumbersomely set up a server to host the data. Reference the [previous module](https://ploomber-sql.readthedocs.io/en/latest/packaging-your-sql-project/intro-to-python-scripting-and-pipelines.html) to learn more about using Python with DuckDB. 
+
+2. `pandas` is an extensive package. Comprehensively going through the entire library's documentation would be time-consuming and unnecessary for your understanding of the script. The script relies only on typical `pandas` functions, such as ones that help rename columns appropriately, clean data for null values, and create new columns. 
+
+Let's now dissect `datadownload.py` and focus specifically on key lines of its ETL process.
+
+![](etl-co2-sample.jpg)
+
+### Extracting with Python
+
+Take a look at the function `fuel_consumption_metadata_extraction`. On lines 100 - 115, we access the Government of Canada's API and extract data regarding fuel consumption ratings that are in English. The output of this function is a `pandas` data frame with two columns: "name" and "url". Each row in the data frame corresponds to a single data set under the fuel consumptions ratings portfolio that is in English.  
+
+```python
```

Comment:
In the previous section we covered why we develop functions and why they form an important part of the scripting process, it would look cleaner if you share the full function and explain its different key components:

1. Try-except clauses: in this case they help us catch the case when the API returns an error
2. Transform the json response into pandas format
3. Select entries in english
4. Identify the paths to the full data

Reply:
The section now focuses more so on the function rather than the main program.

Reply:
Awesome!